### DISCUSSION TITLE
Update system from one plugin running before Startup system of another

### DISCUSSION THREAD

**Discussion by FelipeTJ** [bevyengine/bevy#11833] [Q&A] [answered]
I have a couple different plugins running in my game and I want to run a few setup systems on startup and others every frame. I don't know if the details matter but I am setting up a camera as a child of a Scene (which is the player's 3d character model). The main parts of the code I use are this:

To load and setup the heirarchy:
```rust
pub struct GameplayPlugin;

impl Plugin for GameplayPlugin {
    fn build(&self, app: &mut App) {
        app
            .add_plugins((MovementPlugin, CameraPlugin))
            .add_systems(PreStartup, player_asset_loader)
            .add_systems(Startup, spawn_player)
            .add_systems(PostStartup, setup_camera_child);
    }
}
```
To handle camera movement:
```rust
pub struct CameraPlugin;

impl Plugin for CameraPlugin {
    fn build(&self, app: &mut App) {
        app
            .add_systems(Update, change_camera)
            .add_systems(Update, transition_camera.run_if(camera_is_transitioning));  
    }
}
```
The main
```rust
fn main() {
    App::new()
    .insert_resource(ClearColor(Color::rgb(0.2, 0.2, 0.2)))
    .add_plugins(DefaultPlugins.set(WindowPlugin {
        primary_window: Some(Window {
            title: "Obs game test".to_string(),
            resizable: false,
            // Turn off vsync to maximize CPU/GPU usage
            present_mode: PresentMode::AutoVsync,
            // present_mode: PresentMode::AutoNoVsync,
            ..default()
        }),
        ..default()
    }))
    .add_plugins(WorldInspectorPlugin::new())
    .add_plugins((GameplayPlugin, WorldPlugin))
    .add_systems(Update, bevy::window::close_on_esc)
    .run();
}
```

However, I get this error:
```rust
thread 'Compute Task Pool (0)' panicked at src/player_mechanics/camera.rs:163:22:
called `Result::unwrap()` on an `Err` value: NoEntities("bevy_ecs::query::state::QueryState<&obs_game::player_mechanics::camera::CameraSettings>")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Encountered a panic in system `bevy_app::main_schedule::Main::run_main`!
```

That comes from:
```rust
fn camera_is_transitioning(q: Query<&CameraSettings>) -> bool {
    let settings = q.single();
    settings.target_view_state.is_some()
}
```

This leads me to believe that the system scheduled for `Update` is running before `PostStartup`. Can this be the error? If so, is it a bug or could it be something else from my code?

**Comment by pablo-lua** [comment, marked as answer]
The line at fault is probably 
```rust
    let settings = q.single();
```
can panic if you don't have spawned the Camera yet

you can solve it by 
- Making sure that the entity is spawned before anything (I'm not sure if this will solve the problem, as this is a run condition and maybe can happen at some other time?)
- get_single instead of single ->
```rust
   let Ok(settings) = q.get_single() else {
      return false;
   }
   // Your logic here
```

**Reply by FelipeTJ** [reply]
Thanks! I added that as a filter for running those functions and it works properly now.

I wonder if there is a better way to prevent any overhead from systems that constantly query but simply return immediately... Or maybe the overhead is not too bad?

**Reply by hymm** [reply]
Query overhead is not that bad. The main cost to them is updating the matching archetypes, which is not that expensive. Also this only has a cost when there are new archetypes.
